nomad/secure_variables_endpoint.go

```diff
@@ -1,133 +1,470 @@
+			return mErr.ErrorOrNil()
+		}
+
+		// At this point, we do not have a valid ACL token, nor are we being
+		// called, or able to confirm via the state store, by a node.
+		if node == nil {
+			return structs.ErrTokenNotFound
+		}
 	}
 
-	sv.logger.Trace("TODO") // silences structcheck lint
+	// Use a multierror, so we can capture all validation errors and pass this
+	// back so fixing in a single swoop.
+	var mErr multierror.Error
 
-	// TODO: placeholder for serialization and encryption
-	var buf bytes.Buffer
-	enc := gob.NewEncoder(&buf)
-	err := enc.Encode(args.Data.UnencryptedData)
-	if err != nil {
+	// Iterate the secure variables and validate them. Any error results in the
+	// call failing.
+	for _, i := range args.Data {
+		if err := i.Validate(); err != nil {
+			mErr.Errors = append(mErr.Errors, err)
+		}
+		if err := sv.encrypt(i); err != nil {
+			mErr.Errors = append(mErr.Errors, err)
+		}
+	}
+	if err := mErr.ErrorOrNil(); err != nil {
 		return err
 	}
 
-	args.Data.EncryptedData = &structs.SecureVariableData{}
-	args.Data.EncryptedData.KeyID = "TODO"
-	args.Data.EncryptedData.Data = sv.encrypter.Encrypt(buf.Bytes(), args.Data.EncryptedData.KeyID)
+	// Update via Raft.
+	out, index, err := sv.srv.raftApply(structs.SecureVariableUpsertRequestType, args)
+	if err != nil {
+		return err
+	}
 
-	// TODO: implementation
-	SV_Upsert(args, reply)
+	// Check if the FSM response, which is an interface, contains an error.
+	if err, ok := out.(error); ok && err != nil {
+		return err
+	}
 
+	// Update the index. There is no need to floor this as we are writing to
+	// state and therefore will get a non-zero index response.
+	reply.Index = index
 	return nil
 }
 
-func (sv *SecureVariables) List(args *structs.SecureVariablesListRequest, reply *structs.SecureVariablesListResponse) error {
-	if done, err := sv.srv.forward("SecureVariables.List", args, args, reply); done {
+// Delete removes a single secure variable, as specified by its namespace and
+// path from Nomad. This is typically called by Nomad nodes, however, in extreme
+// situations can be used via the CLI and API by operators.
+func (sv *SecureVariables) Delete(
+	args *structs.SecureVariablesDeleteRequest,
+	reply *structs.SecureVariablesDeleteResponse) error {
+
+	if done, err := sv.srv.forward(structs.SecureVariablesDeleteRPCMethod, args, args, reply); done {
 		return err
 	}
+	defer metrics.MeasureSince([]string{"nomad", "secure_variables", "delete"}, time.Now())
 
-	defer metrics.MeasureSince([]string{"nomad", "secure_variables", "list"}, time.Now())
+	// Perform the ACL token resolution.
+	aclObj, err := sv.srv.ResolveToken(args.AuthToken)
+
+	switch err {
+	case nil:
+		// If ACLs are enabled, ensure the caller has the submit-job namespace
+		// capability.
+		if aclObj != nil {
+			hasSubmitJob := aclObj.AllowNsOp(args.RequestNamespace(), acl.NamespaceCapabilitySubmitJob)
+			if !hasSubmitJob {
+				return structs.ErrPermissionDenied
+			}
```

Comment:
Obviously we don't have the new ACL policy set up yet but we should leave a TODO here just for ourselves. 😀 

Same applies to `SecureVariables.Upsert`